htlcswitch/link.go

```diff
@@ -1473,6 +1489,41 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 		}
 
 	case *lnwire.CommitSig:
+		// Since we may have learned new preimages for the first time,
+		// we'll add them to our preimage cache. By doing this, we
+		// ensure any contested contracts watched by any on-chain
+		// arbitrators can now sweep this HTLC on-chain. We delay
+		// committing the preimages until just before accepting the new
+		// remote commitment, as afterwards the peer won't resend the
+		// Settle messages on the next channel reestablishment. Doing so
+		// allows us to more effectively batch this operation, instead
+		// of doing a single write per preimage.
+		if len(l.uncommitedPreimages) > 0 {
```

Comment:
Is it right that this performance improvement is only observable if multiple htlcs are fulfilled before a new commitsig is sent? So on a very high frequency link?

Reply:
That's correct, yes

Reply:
Ok. But the main reason for this PR is the unsafety of writing the preimage?

Reply:
The 80% penalty comes when writing each preimage individually, using our async_payments_benchmark integration test as a yard stick (which does attempt to saturate the links). For the single payment case, there would be a negligible penalty since the write is now done in the critical path

Reply:
Ok, got it.